### PR TITLE
Require end date to launch

### PR DESCRIPTION
**Pull request by jessie-ross**
**This PR adds a required end date to launch an experiment:**

Screenshot 2023-05-17 at 4 37 41 pm

Due to technical limitations, I update the status and end date separately, first the status and then the end date. Ideally we'd be updating the end date first but we don't have a patch endpoint until the experiment is launched, and even if we could we can't update both the status and end date at the same time as updating the status requires a separate endpoint anyway.

This results in some subtle fragility as the two updates that should occur at the same time are happening sequentially, and means if there is an error with the end date request, we could have experiments launched without an end date.

This should be an acceptable risk as it isn't the end of the world if an experiment gets launched without an end date.

Part of https://github.com/Automattic/experimentation-platform/issues/875

## How has this been tested?

- Automated tests that cover added/changed functionality
- Manual testing with production data (locally)

### CODE REVIEW

**Review by aaronyan — COMMENTED** [review]
Looks great, thank you very much for adding this, @jessie-ross.

> Due to technical limitations, I update the status and end date separately, first the status and then the end date.

Very interesting technical limitation. Thanks for documenting! Can you create an issue sharing the improvements you would recommend to fix these technical limitations?

> adds a required end date to launch an experiment

I prefer to add duration days as the input but am good with having an end date here. What was your reasoning for having the input be a specific end date?

> Danger

In a future iteration, I think we can consider softening this design. Manual experiment start will become the only option for launching an experiment, and I would prefer reserving the 'danger' design for special and uncommon scenarios (like disabling a running experiment). The 'danger' design would be part of the expected experiment cycle, which seems odd to me. 

I've created an issue for now (https://github.com/Automattic/experimentation-platform/issues/952) and we can discuss there more about whether we should soften this design language.

File: src/components/explat/experiments/single-view/ExperimentRunButton.tsx [hunk 1]
```diff
@@ -110,21 +161,52 @@ const ExperimentRunButton = ({
...
+          initialValues={{ experiment: editInitialExperiment }}
+          validationSchema={yup.object({ experiment: editValidationSchema }).required()}
+          validateOnMount
+          onSubmit={onSubmit}
+        >
+          {(formikProps) => (
+            <form onSubmit={formikProps.handleSubmit} noValidate>
+              <DialogContent>
+                <Typography variant='body2' gutterBottom>
+                  An eventual end date is <strong>required</strong> to prevent experiments running forever:
```
[the remark this answers is not in the document]
Comment (jessie-ross) [reply src/components/explat/experiments/single-view/ExperimentRunButton.tsx:174]: I think eventual is the right phrasing, this end date is more of a safeguard - this experiment will be completed by this date - rather then being used as automated completion.

**Comment by jessie-ross** [comment]
> Very interesting technical limitation. Thanks for documenting! Can you create an issue sharing the improvements you would recommend to fix these technical limitations?

I recommend no changes and leaving it as is.

> I prefer to add duration days as the input but am good with having an end date here. What was your reasoning for having the input be a specific end date?

Yeah I don't see too much difference between the two approaches, but this is a bit simpler code-wise.

**Comment by aaronyan** [comment]
> Yeah I don't see too much difference between the two approaches, but this is a bit simpler code-wise.

Thanks for sharing! I think it's easier for the experimenter to enter the number of days instead of the exact date. Created an issue for a future enhancement! https://github.com/Automattic/experimentation-platform/issues/955

**Comment by aaronyan** [comment]
> Due to technical limitations, I update the status and end date separately, first the status and then the end date

@jessie-ross, I thought we can update an experiment in `staging` using `PUT` (`PATCH` is for non-staging experiments, see `class-experiments-controller.php`). Would this work with the `end_datetime` before transitioning the status to `running`?

**Comment by jessie-ross** [comment]
> @jessie-ross, I thought we can update an experiment in staging using PUT (PATCH is for non-staging experiments, see class-experiments-controller.php). Would this work with the end_datetime before transitioning the status to running?

We can, but it is a bit more complicated because a PUT request needs to have all the experiment data sent through.
IMO the current solution won't cause any problems.

**Pull request merged**